presto-spark-base/src/main/java/com/facebook/presto/spark/execution/http/PrestoSparkHttpWorkerClient.java

```diff
@@ -199,8 +220,79 @@ public ListenableFuture<BaseResponse<TaskInfo>> updateTask(
                 createAdaptingJsonResponseHandler(taskInfoCodec));
     }
 
+    public ListenableFuture<BaseResponse<ServerInfo>> getServerInfo()
+    {
+        Request request = setContentTypeHeaders(false, prepareGet())
+                .setUri(serverUri)
+                .build();
+        return httpClient.executeAsync(request, createAdaptingJsonResponseHandler(serverInfoCodec));
+    }
+
+    public SettableFuture<ServerInfo> getServerInfoWithRetry()
+    {
+        SettableFuture<ServerInfo> future = SettableFuture.create();
+        doGetServerInfo(future);
+        return future;
+    }
+
     public URI getLocation()
     {
         return location;
     }
+
+    private void doGetServerInfo(SettableFuture<ServerInfo> future)
+    {
+        addCallback(getServerInfo(), new FutureCallback<BaseResponse<ServerInfo>>()
+        {
+            @Override
+            public void onSuccess(@Nullable BaseResponse<ServerInfo> response)
+            {
+                if (response.getStatusCode() != OK.code()) {
+                    throw new PrestoException(GENERIC_INTERNAL_ERROR, "Request failed with HTTP status " + response.getStatusCode());
+                }
+                future.set(response.getValue());
+            }
+
+            @Override
+            public void onFailure(Throwable failedReason)
+            {
+                if (failedReason instanceof RejectedExecutionException && httpClient.isClosed()) {
+                    log.error("Unable to start the native process. HTTP client is closed");
+                    future.setException(failedReason);
+                    return;
+                }
+                // record failure
+                try {
+                    errorTracker.requestFailed(failedReason);
```

Comment:
If we choose to use the RequestErrorTracker to handle failure, it will be nice to also change result fetch and info fetch to use the same, to be consistent.

Reply:
Or we probably won't be needing that fancy backoff logic in the tracker at all. That IMO works better in a multi-tenant environment when resource contention is a commonly happening scenario. It could just be simple logic of n times retry with interval of m seconds. Up to you.